### PR TITLE
add wam support

### PR DESCRIPTION
**Pull request by xiangyan99**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by xiangyan99** [comment]
#23687

File: sdk/identity/azure-identity/azure/identity/_credentials/certificate.py [hunk 1]
```diff
@@ -45,10 +45,22 @@ class CertificateCredential(ClientCredentialBase):
     :keyword cache_persistence_options: configuration for persistent token caching. If unspecified, the credential
         will cache tokens in memory.
     :paramtype cache_persistence_options: ~azure.identity.TokenCachePersistenceOptions
+    :keyword bool allow_broker:  
+        Brokers provide Single-Sign-On, device identification,
+        and application identification verification.
+        If this parameter is set to True,
+        the broker will be used when possible. Default to False.
```
Comment (mccoyp) [inline sdk/identity/azure-identity/azure/identity/_credentials/certificate.py:52]: Are the line breaks here intentional? The following has the same line breaking as the rest of the docstring:
```suggestion
    :keyword bool allow_broker: Brokers provide single sign-on, device identification, and application identification
        verification. If this parameter is set to True, the broker will be used when possible. Defaults to False.
```

File: sdk/identity/azure-identity/azure/identity/_credentials/certificate.py [hunk 2]
```diff
@@ -45,10 +45,22 @@ class CertificateCredential(ClientCredentialBase):
...
 
-    def __init__(self, tenant_id, client_id, certificate_path=None, **kwargs):
-        # type: (str, str, Optional[str], **Any) -> None
+    def __init__(
+            self,
+            tenant_id: str,
+            client_id: str,
+            certificate_path: str = None,
+            *,
+            allow_broker: bool = False,
```
Comment (mccoyp) [inline sdk/identity/azure-identity/azure/identity/_credentials/certificate.py:61]: Accepting `allow_broker` as a named keyword-only argument pulls it out of `kwargs`, so I think it'll need to be explicitly passed to the `super().__init__` call below

File: sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 if TYPE_CHECKING:
     # pylint:disable=unused-import,ungrouped-imports
-    from typing import Any
+    from typing import Any, Optional
```
Comment (mccoyp) [inline sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py:11]: It looks like neither of these imports are used

File: sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py [hunk 4]
```diff
@@ -24,10 +24,22 @@ class ClientSecretCredential(ClientCredentialBase):
     :keyword cache_persistence_options: configuration for persistent token caching. If unspecified, the credential
         will cache tokens in memory.
     :paramtype cache_persistence_options: ~azure.identity.TokenCachePersistenceOptions
+    :keyword bool allow_broker:
```
Comment (mccoyp) [inline sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py:27]: Same comments about the line break formatting and explicit passing of `allow_broker` to the parent class

File: sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py [hunk 5]
```diff
@@ -48,10 +48,21 @@ class OnBehalfOfCredential(MsalCredential, GetTokenMixin):
         is a unicode string, it will be encoded as UTF-8. If the certificate requires a different encoding, pass
         appropriately encoded bytes instead.
     :paramtype password: str or bytes
+    :keyword bool allow_broker:
```
Comment (mccoyp) [inline sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py:51]: Same comments as for other classes

**Review by mccoyp — COMMENTED** [review]
Are there any tests we can add for this? At least for verifying that the kwarg gets passed along correctly?

File: sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py [hunk 6]
```diff
@@ -24,10 +20,19 @@ class ClientSecretCredential(ClientCredentialBase):
...
 
-    def __init__(self, tenant_id, client_id, client_secret, **kwargs):
-        # type: (str, str, str, **Any) -> None
+    def __init__(
+            self,
+            tenant_id: str,
+            client_id: str,
+            client_secret: str,
+            *,
+            allow_broker: bool = False,
```
Comment (mccoyp) [inline sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py:33]: Is `allow_broker` a term used in other libraries (`msal`, Identity in other languages, etc.)? Has the team agreed on this, or do you think something like `use_broker` would make more sense? I'm not familiar with the feature, so I'll defer to you though 🙂

Comment (xiangyan99) [reply sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py:33]: It is used in msal.

This is a beta feature and the parameter name is subject to change when doing arch board review. :) (btw, it is unlikely to GA in copper)

Comment (mccoyp) [reply sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py:33]: Got it, thanks!

**Review by mccoyp — APPROVED** [review]

File: sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py [hunk 7]
```diff
@@ -24,10 +18,19 @@ class ClientSecretCredential(ClientCredentialBase):
     :keyword cache_persistence_options: configuration for persistent token caching. If unspecified, the credential
         will cache tokens in memory.
     :paramtype cache_persistence_options: ~azure.identity.TokenCachePersistenceOptions
+    :keyword bool allow_broker: Brokers provide single sign-on, device identification, and application identification
```
Comment (rayluo) [inline sdk/identity/azure-identity/azure/identity/_credentials/client_secret.py:21]: Broker/WAM supports only public client. We do not need to change this file.

File: sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py [hunk 8]
```diff
@@ -48,10 +48,18 @@ class OnBehalfOfCredential(MsalCredential, GetTokenMixin):
         is a unicode string, it will be encoded as UTF-8. If the certificate requires a different encoding, pass
         appropriately encoded bytes instead.
     :paramtype password: str or bytes
+    :keyword bool allow_broker: Brokers provide single sign-on, device identification, and application identification
```
Comment (rayluo) [inline sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py:51]: Broker/WAM supports only public client. We do not need to change this file.

File: sdk/identity/azure-identity/azure/identity/_credentials/certificate.py [hunk 9]
```diff
@@ -45,16 +45,29 @@ class CertificateCredential(ClientCredentialBase):
     :keyword cache_persistence_options: configuration for persistent token caching. If unspecified, the credential
         will cache tokens in memory.
     :paramtype cache_persistence_options: ~azure.identity.TokenCachePersistenceOptions
+    :keyword bool allow_broker: Brokers provide single sign-on, device identification, and application identification
```
Comment (rayluo) [inline sdk/identity/azure-identity/azure/identity/_credentials/certificate.py:48]: Broker/WAM supports only public client. We do not need to change this file.

**Review by pvaneck — COMMENTED** [review]
Is there any documentation or a sample about usage of the `allow_broker` arg we can provide for users?

File: sdk/identity/azure-identity/azure/identity/_credentials/browser.py [hunk 10]
```diff
@@ -42,6 +42,10 @@ class InteractiveBrowserCredential(InteractiveCredential):
         will cache tokens in memory.
     :paramtype cache_persistence_options: ~azure.identity.TokenCachePersistenceOptions
     :keyword int timeout: seconds to wait for the user to complete authentication. Defaults to 300 (5 minutes).
+    :keyword bool allow_broker: Brokers provide single sign-on, device identification, and application identification
+        verification. If this parameter is set to True, the broker will be used when possible. Defaults to False.
+        Check https://learn.microsoft.com/en-us/azure/active-directory/develop/scenario-desktop-acquire-token-wam
```
Comment (pvaneck) [inline sdk/identity/azure-identity/azure/identity/_credentials/browser.py:47]:
```suggestion
        Check https://learn.microsoft.com/azure/active-directory/develop/scenario-desktop-acquire-token-wam
```

File: sdk/identity/azure-identity/azure/identity/_credentials/user_password.py [hunk 11]
```diff
@@ -34,6 +34,10 @@ class UsernamePasswordCredential(InteractiveCredential):
     :keyword cache_persistence_options: Configuration for persistent token caching. If unspecified, the credential
         will cache tokens in memory.
     :paramtype cache_persistence_options: ~azure.identity.TokenCachePersistenceOptions
+    :keyword bool allow_broker: Brokers provide single sign-on, device identification, and application identification
+        verification. If this parameter is set to True, the broker will be used when possible. Defaults to False.
+        Check https://learn.microsoft.com/en-us/azure/active-directory/develop/scenario-desktop-acquire-token-wam
```
Comment (pvaneck) [inline sdk/identity/azure-identity/azure/identity/_credentials/user_password.py:39]:
```suggestion
        Check https://learn.microsoft.com/azure/active-directory/develop/scenario-desktop-acquire-token-wam
```

File: sdk/identity/azure-identity/setup.py [hunk 12]
```diff
@@ -60,7 +60,7 @@
     install_requires=[
         "azure-core<2.0.0,>=1.11.0",
         "cryptography>=2.5",
-        "msal<2.0.0,>=1.12.0",
+        "msal<2.0.0,>=1.20.0b1",
```
Comment (pvaneck) [inline sdk/identity/azure-identity/setup.py:63]: Should we use [1.20.0](https://pypi.org/project/msal/1.20.0/)?

**Review by pvaneck — COMMENTED** [review]
I see that the param was removed from DeviceCodeCredential doc string. I take it WAM is not supported with DeviceCodeCredential?

File: sdk/identity/azure-identity/setup.py (hunk 12 above)
Comment (xiangyan99) [reply sdk/identity/azure-identity/setup.py:63]: > Should we use [1.20.0](https://pypi.org/project/msal/1.20.0/)?

You are absolutely right!

Comment (xiangyan99) [inline sdk/identity/azure-identity/setup.py:63]:
```suggestion
        "msal<2.0.0,>=1.20.0",
```

File: shared_requirements.txt [hunk 13]
```diff
@@ -113,7 +113,7 @@ futures
 mock
 typing
 typing-extensions
-msal<2.0.0,>=1.12.0
+msal<2.0.0,>=1.20.0b1
```
Comment (xiangyan99) [inline shared_requirements.txt:116]:
```suggestion
msal<2.0.0,>=1.20.0
```

**Review by pvaneck — APPROVED** [review]

File: sdk/identity/azure-identity/azure/identity/_credentials/browser.py [hunk 14]
```diff
@@ -51,6 +51,8 @@ class InteractiveBrowserCredential(InteractiveCredential):
         will cache tokens in memory.
     :paramtype cache_persistence_options: ~azure.identity.TokenCachePersistenceOptions
     :keyword int timeout: seconds to wait for the user to complete authentication. Defaults to 300 (5 minutes).
+    :keyword bool allow_broker: Brokers provide single sign-on, device identification, and application identification
+        verification. If this parameter is set to True, the broker will be used when possible. Defaults to False.
```
Comment (rayluo) [inline sdk/identity/azure-identity/azure/identity/_credentials/browser.py:46]: FYI: MSAL teams are currently under discussion on whether this parameter shall be changed to be platform specific, meaning downstream callers would need to opt in only on supported platform(s). Azure Identity shall mark this release a beta release, until we settle down on one way or another.

File: sdk/identity/azure-identity/setup.py [hunk 15]
```diff
@@ -60,7 +60,7 @@
     install_requires=[
         "azure-core<2.0.0,>=1.11.0",
         "cryptography>=2.5",
-        "msal<2.0.0,>=1.12.0",
+        "msal<2.0.0,>=1.20.0",
```
Comment (rayluo) [inline sdk/identity/azure-identity/setup.py:63]: There is a tricky thing here. [MSAL Python chose to declare the broker (i.e. WAM in this context) dependency as extra](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.20.0/setup.py#L89-L96).

Azure Identity may choose to also expose an [extra functionality](https://peps.python.org/pep-0508/#extras) as `azure-identity[broker]` and then leave the decision for app developers. Alternatively, Azure Identity can simply always pull in MSAL's broker bits by `msal[broker]>=1.20.0,<2`.

File: shared_requirements.txt [hunk 16]
```diff
@@ -113,7 +113,7 @@ futures
 mock
 typing
 typing-extensions
-msal<2.0.0,>=1.12.0
+msal<2.0.0,>=1.20.0
```
Comment (rayluo) [inline shared_requirements.txt:116]: See above.

**Pull request merged**